BUG: Repeated calls to the var method of complex MaskedArrays lead to increasing large results

### Describe the issue:

On windows10 using the main branch  50984037 (open):Merge pull request #23878 from charris/backport-23846 repeated calls to the var method of a complex MaskedArrays lead to increasingly large results.

An old version on linux does not show the problem:
1.20.3
3.8.10 (default, May 26 2023, 14:05:08) 
[GCC 9.4.0]


​

### Reproduce the code example:

```python
import numpy.ma as ma
import numpy as np

# Base data definition.
x = np.array([8.375j, 7.545j, 8.828j, 8.5j, 1.757j, 5.928,
              8.43, 7.78, 9.865, 5.878, 8.979, 4.732,
              3.012, 6.022, 5.095, 3.116, 5.238, 3.957,
              6.04, 9.63, 7.712, 3.382, 4.489, 6.479j,
              7.189j, 9.645, 5.395, 4.961, 9.894, 2.893,
              7.357, 9.828, 6.272, 3.758, 6.693, 0.993j])

mask = np.array([0, 1, 0, 1, 0, 0,
             1, 0, 1, 1, 0, 1,
             0, 0, 0, 1, 0, 1,
             0, 0, 0, 1, 1, 1,
             1, 0, 0, 1, 0, 0,
             0, 0, 1, 0, 1, 0], dtype=bool)

mx = ma.array(data=x, mask=mask)

mXcomp = mx.compressed()

for i in range(10):
    # print(np.min(mXcomp),np.max(mXcomp))
    #print(mXcomp)
    print(mXcomp.var(), x.var())


'''Python
Example output:
16.211890209804558 19.282807968364196
16.257534793652272 19.282807968364196
16.390019533407436 19.282807968364196
17.271411574130035 19.282807968364196
41.913043714992426 19.282807968364196
13938.461926153328 19.282807968364196
4070437439.6019826 19.282807968364196
3.479376771804534e+20 19.282807968364196
2.5422731712363174e+42 19.282807968364196
1.3572621042094718e+86 19.282807968364196
'''
```


### Error message:

_No response_

### Runtime information:

1.25.0rc1+90.g50984037f
3.10.11 | packaged by Anaconda, Inc. | (main, Apr 20 2023, 18:56:50) [MSC v.1916 64 bit (AMD64)]

### Context for the issue:

It is a likely pointer/memory leakage problem, and it also affects the outcomes of other computations using the same matrix.

It might be related to the reason some tests for numpy.ma fail:

python runtests.py -t numpy/ma/tests/
is failing with weird errors on complex masked arrays, particularly:
 test_varstd
 test_complex

Sorry for the delayed response, I can confirm its definitely a compiler bug since msvc v19.30, and it can also be trrigerd via `test_simd.py`:
```Bash

self = <numpy.core.tests.test_simd.Test_SIMD_ALL_256_AVX2_f64 object at 0x0000022A2FE26FD0>, intrin = 'self.store2_till', elsizes = (32, 64), scale = 2

    @pytest.mark.parametrize("intrin, elsizes, scale", [
        ("self.store_till", (32, 64), 1),
        ("self.store2_till", (32, 64), 2),
    ])
    def test_memory_partial_store(self, intrin, elsizes, scale):
        if self._scalar_size() not in elsizes:
            return
        npyv_store_till = eval(intrin)
        data = self._data()
        data_rev = self._data(reverse=True)
        vdata = self.load(data)
        lanes = list(range(1, self.nlanes + 1))
        lanes += [self.nlanes**2, self.nlanes**4]
        for n in lanes:
            data_till = data_rev.copy()
            data_till[:n*scale] = data[:n*scale]
            store_till = self._data(reverse=True)
            npyv_store_till(store_till, n, vdata)
>           assert store_till == data_till
E           assert [4.0, 3.0, 2.0, 1.0] == [1.0, 2.0, 2.0, 1.0]
E             At index 0 diff: 4.0 != 1.0
E             Full diff:
E             - [1.0, 2.0, 2.0, 1.0]
E             + [4.0, 3.0, 2.0, 1.0]

data       = [1.0, 2.0, 3.0, 4.0]
data_rev   = [4.0, 3.0, 2.0, 1.0]
data_till  = [1.0, 2.0, 2.0, 1.0]
elsizes    = (32, 64)
intrin     = 'self.store2_till'
lanes      = [1, 2, 3, 4, 16, 256]
n          = 1
npyv_store_till = <built-in function store2_till_f64>
scale      = 2
self       = <numpy.core.tests.test_simd.Test_SIMD_ALL_256_AVX2_f64 object at 0x0000022A2FE26FD0>
store_till = [4.0, 3.0, 2.0, 1.0]
vdata      = <npyv_f64 of [1.0, 2.0, 3.0, 4.0]>

..\..\..\AppData\Roaming\Python\Python311\site-packages\numpy-1.25.0rc1+189.gff5c1bf38-py3.11-win-amd64.egg\numpy\core\tests\test_simd.py:727: AssertionError
=========================================================================================== short test summary info ============================================================================================
FAILED ..\..\core\tests\test_simd.py::Test_SIMD_ALL_256_AVX2_u64::test_memory_partial_store[self.store2_till-elsizes1-2] - assert [4, 3, 2, 1] == [1, 2, 2, 1]
FAILED ..\..\core\tests\test_simd.py::Test_SIMD_ALL_256_AVX2_s64::test_memory_partial_store[self.store2_till-elsizes1-2] - assert [4, 3, 2, 1] == [1, 2, 2, 1]
FAILED ..\..\core\tests\test_simd.py::Test_SIMD_ALL_256_AVX2_f64::test_memory_partial_store[self.store2_till-elsizes1-2] - assert [4.0, 3.0, 2.0, 1.0] == [1.0, 2.0, 2.0, 1.0]
```

Only AVX2 is affected by this bug, so disabling `AVX2` via `NPY_DISABLE_CPU_FEATURES` should bypass it.

Here a proof of the bug https://godbolt.org/z/4GPP4KT4d, simply the compiler optimizer performs 256-bit mov rather than 128-bit when the pair length is `1`:
```ASM
npyv_store2_till_s64 PROC                         ; COMDAT
        vmovdqu ymm0, YMMWORD PTR [r8]
        cmp     rdx, 1
        jbe     SHORT $LN6@npyv_store
        vmovdqu YMMWORD PTR [rcx], ymm0
$LN6@npyv_store:
        vzeroupper
        ret     0
```
Expected to be:

```Asm
        vmovdqu ymm0, YMMWORD PTR [r8]
        vmovdqu XMMWORD PTR [rcx], xmm0
        cmp     rdx, 1
        jbe     SHORT $LN6@npyv_store
        vextracti128 XMMWORD PTR [rcx+16], ymm0, 1
$LN6@npyv_store:
```

--- a/numpy/core/src/common/simd/avx2/memory.h
+++ b/numpy/core/src/common/simd/avx2/memory.h
@@ -372,10 +372,29 @@ NPY_FINLINE void npyv_store2_till_s32(npy_int32 *ptr, npy_uintp nlane, npyv_s32 
 NPY_FINLINE void npyv_store2_till_s64(npy_int64 *ptr, npy_uintp nlane, npyv_s64 a)
 {
     assert(nlane > 0);
+#ifdef _MSC_VER
+   /*
+    * Although this version is compatible with all other compilers,
+    * there is no performance benefit in retaining the other branch.
+    * However, it serves as evidence of a newly emerging bug in MSVC
+    * that started to appear since v19.30.
+    * For some reason, the MSVC optimizer chooses to ignore the lower store (128-bit mov)
+    * and replace with full mov counting on ymmword pointer.
+    *
+    * For more details, please refer to the discussion on https://github.com/numpy/numpy/issues/23896.
+    */
+    if (nlane > 1) {
+        npyv_store_s64(ptr, a);
+    }
+    else {
+        npyv_storel_s64(ptr, a);
+    }
+#else
     npyv_storel_s64(ptr, a);
     if (nlane > 1) {
         npyv_storeh_s64(ptr + 2, a);
     }
+#endif
 }
 /*********************************
  * Non-contiguous partial store

